src/app/state/importaccounts/saga.ts

```diff
@@ -71,33 +78,64 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
   }
 }
 
+function* fetchBalanceForAccount(account: ImportAccountsListAccount) {
+  let currentStep = (yield* select(selectImportAccounts)).step
+  if (currentStep === ImportAccountsStep.Idle) {
+    yield* setStep(ImportAccountsStep.LoadingBalances)
+  }
+  const balance = yield* call(getBalance, hex2uint(account.publicKey))
+  yield* put(
+    importAccountsActions.updateAccountBalance({
+      address: account.address,
+      balance,
+    }),
+  )
+  currentStep = (yield* select(selectImportAccounts)).step
+  const hasMissingBalances = yield* select(selectImportAccountHasMissingBalances)
+  if (currentStep === ImportAccountsStep.LoadingBalances && !hasMissingBalances) {
+    yield* setStep(ImportAccountsStep.Idle)
+  }
+}
+
+function* ensureAllBalancesArePresentOnCurrentPage() {
+  const existingAccounts = yield* select(selectImportAccountsOnCurrentPage)
+  const accountsWithNoBalance = existingAccounts.filter(a => !a.balance)
+  for (let i = 0; i < accountsWithNoBalance.length; i++) {
+    yield* fork(fetchBalanceForAccount, accountsWithNoBalance[i])
+  }
+}
+
+/**
+ * Enumerate more accounts from Ledger, enough to fill up one page.
+ */
 function* enumerateAccountsFromLedger() {
+  const existingAccounts = yield* select(selectImportAccountsFullList)
+  const pageNumber = yield* select(selectImportAccountsPageNumber)
+  if (existingAccounts.length >= (pageNumber + 1) * accountsPerPage) return
   yield* setStep(ImportAccountsStep.OpeningUSB)
   let transport: Transport | undefined
   try {
     transport = yield* getUSBTransport()
+    const existingAccounts = yield* select(selectImportAccountsFullList)
+    const start = existingAccounts.length
 
     yield* setStep(ImportAccountsStep.LoadingAccounts)
-    const accounts = yield* call(Ledger.enumerateAccounts, transport)
+    for (let index = start; index < start + accountsPerPage; index++) {
+      const account = (yield* call(Ledger.enumerateAccounts, transport, 1, index))[0]
```

Comment:
This is odd. Refactor enumerateAccounts?

And add a ledger-specific accountsPerPage

Reply:
> Refactor enumerateAccounts?

Refactored.

> And add a ledger-specific accountsPerPage

Why should it be ledger-specific? (Ledger is not the only limiting factor; we are constrained to 4 acc per page because of the extension sizing issues anyway.) Do we want to introduce different rules for all small screen / big screen / mnemonic / ledger combinations?

Reply:
I thought we might be increasing accountsPerPage for mnemonic at some point, and forget about ledger performance issues.

With new decrease, let's add a comment that accountsPerPage needs to be small enough for extension layout and small enough for slow ledgers

Reply:
OK, comment added.